### PR TITLE
Decompile ABI using a heimdall-rs backend

### PR DESCRIPTION
**Pull request by portdeveloper**
## Description

This PR:

- Calls [a simple heimdall-rs backend](https://github.com/portdeveloper/gulltoppr) to get the abi of a contract 
- Implements steps instead of tabs in the ui
- Enriches notifications to provide better UX.

https://github.com/BuidlGuidl/abi.ninja/assets/108868128/a1d7e508-6607-47ad-a431-0fd2ea209196

### ToDo - Improvements
- [x] go back button looks bad on smaller viewports, also its placement can be better
- [ ] code cleanup

## Additional Information

- [ ] This is not a duplicate of any [existing pull request](https://github.com/BuidlGuidl/abi.ninja/pulls)
- [ ] I have read the [contributing docs](https://github.com/BuidlGuidl/abi.ninja/blob/main/CONTRIBUTING.md) (if this is your first contribution)

## Related Issues

_Closes #63 

_Note: If your changes are small and straightforward, you may skip the creation of an issue beforehand and remove this section. However, for medium-to-large changes, it is recommended to have an open issue for discussion and approval prior to submitting a pull request._

Your ENS/address:

### CODE REVIEW

**Comment by portdeveloper** [comment]
[linkt to the commit](https://github.com/BuidlGuidl/abi.ninja/pull/71/commits/ce5d6e8128d68d87bf3fa1a1783454a98ac9db7a)
The way I made the height of the component here is tricky... I basically used the same div with an empty char instead of the go back button.

https://github.com/BuidlGuidl/abi.ninja/assets/108868128/d051ac39-f19f-4b6f-a53b-f6bd2e1028f0

I am aware this is a bad practice, and will try to fix it asap!

**Comment by Pabl0cks** [comment]
Good job @portdeveloper incorporating heimdall-rs into abi.ninja!! 👾

Last week in London we reviewed the new workflow a bit, and I think we could get some designer help to simplify the UI to make it a bit smoother. Asked Andrea for help and she is going to play a bit in figma to propose a way to simplify it. 

As soon as we get the design I'll send it to review 🙌

**Comment by Pabl0cks** [comment]
Andrea sent us the new UI/UX proposal, let me know what do you think @portdeveloper ! 🙌

The workflow would be something like this:

- User access abi.ninja, we delete the tabs and only leave the contract address input (+ quick access)
- User enters an address and clicks "Load Contract"
   - If contract is verified on the selected network => `ContractDetailPage` loads with the ABI data
   - If contract is not verified, we let the user choose between `Decompyle (beta)` and `Manually import`

We have those design proposals for the "contract not verified" window. 

abi-ninja-Decompile-ManuallyImport

I personally like option 1 and 3, probably 1 is the simplest. I think we can leave network dropdown styles as they are right now for simplicity, don't need to follow the design 100%.

**Comment by portdeveloper** [comment]
https://github.com/BuidlGuidl/abi.ninja/assets/108868128/d4085f90-fcac-4e92-a70e-c17219739065

Updated the UI a little.

**Review by Pabl0cks — CHANGES_REQUESTED** [review]
Nice job @portdeveloper !! With the new UI this PR is looking awesome. Did a first round of testing, these are my notes:

- Check alignment of the header section from the new screen (Heimdall & Import ABI), specially on mobile. "Go back", small logo and network dropdown should be aligned.
- Network dropdown in the new screen should be the same you selected in the homepage search. 
- I think we should disable network dropdown in the new screen, does not make sense to change it there. I see it just informative of what you just searched in the previous screen.
- Remove notification error message when not verified, we already giving all the feedback with the new screen.
- When we access a contract through a URL (/contractAddress/chainId), and the contract is not verified, should we load the new screen instead of showing a not verified error? => If we see this interesting, can create a new issue to tackle it in a future PR, I wouldn't tackle now.

Tomorrow I'll try to review more in depth 🙌.

My test contract in case it's useful for anyone: [sepoliaUnverifiedContract.txt](https://github.com/BuidlGuidl/abi.ninja/files/14797149/sepoliaUnverifiedContract.txt)

**Review by Pabl0cks — APPROVED** [review]
TYSM @portdeveloper for resolving everything so quickly. I've pushed some visual tweaks & fixes.

From a functional standpoint the PR is looking good to me! Would love if @technophile-04 or @carletex could do a technical review 🙌♥

A few comments to check:
- In the new screen we've removed the `NetworksDropdown`, I was thinking in keep showing but disabled, but after reviewing I'm not sure if it's really needed in that step.
- The `MiniFooter` was not clickable, fixed it with ´z-index` but not sure if that's the correct way.
- Added ´MiniFooter´ to the new screen, remove if you feel it doesn't add any value or makes the screen too messy 🙌
- Getting an error when I try to search for an address in Goerli. Maybe it's just a localhost problem.

Thank you again! 🔥

**Comment by carletex** [comment]
> Getting an error when I try to search for an address in Goerli. Maybe it's just a localhost problem.

I think we should remove Goerli from ABI Ninja since it's deprecated. https://blog.ethereum.org/2023/11/30/goerli-lts-update

---

Not sure how far we are from merging, but before we do, should we put Heimdall backend in a BG account? Maybe we can deploy a new instance from Port's [gulltoppr](https://github.com/portdeveloper/gulltoppr) or just adding admin@bg to the existing one. For the sake of paying (if needed), security and not having a single person behind it (as we do with all the BG servers)

Thank you all!!

File: packages/nextjs/components/NetworksDropdown.tsx [hunk 1]
```diff
@@ -56,7 +56,13 @@ const IconOption = (props: OptionProps<Options>) => (
   </Option>
 );
 
-export const NetworksDropdown = ({ onChange }: { onChange: (options: any) => any }) => {
+export const NetworksDropdown = ({
+  onChange,
+  additionalClasses,
+}: {
+  onChange: (options: any) => any;
+  additionalClasses?: string;
```
Comment (technophile-04) [inline packages/nextjs/components/NetworksDropdown.tsx:64]: I think for now lets remove this prop, since we are no more using it

**Review by technophile-04 — APPROVED** [review]
Working great and looks really nice and I feel ready for merge(maybe just after https://github.com/BuidlGuidl/abi.ninja/pull/71#issuecomment-2029536702), Tysm @portdeveloper !!! also thanks @Pabl0cks for fixes and test 🙌

--- 

Some nitpicks for another PR's and discussion : 

> In the new screen we've removed the NetworksDropdown, I was thinking in keep showing but disabled, but after reviewing I'm not sure if it's really needed in that step.

Umm, lol not completely sure either but maybe instead of showing NetworksDropdown disbaled, maybe we could just show plain network name at top right corner (this is might look horrible and ruin design) 

Lol another thing, which might be just me problem but the description above "Decompile (beta)" btn seems the description for only the decompile unitl you read it and notice it serves purpose for whole 2nd screen, maybe we could move it above ? 

<details> 
<summary>

example: 

</summary>

Screenshot 2024-04-01 at 7 29 46 PM


</details>

----

Just some finding : 

This is might be some internal issue with heimdall itself and not in our hands, but for this random contract 
https://etherscan.io/address/0x2e175f748976cd5cdb98f12d1abc5d137d6c9379 its not working

**Comment by portdeveloper** [comment]
@technophile-04 
image
Yes, you are right, heimdall-rs currently cannot decompile that contract and returns with an error, but my backend does not handle it gracefully.
For now I am adding a check for abi array length which returns if the array length is 0.
Tomorrow I will fix the backend to handle this kind of errors better.
Thanks for pointing this out!

**Comment by technophile-04** [comment]
> should we put Heimdall backend in a BG account? Maybe we can deploy a new instance from Port's [gulltoppr](https://github.com/portdeveloper/gulltoppr) or just adding admin@bg to the existing one. For the sake of paying (if needed), security and not having a single person behind it (as we do with all the BG servers)

I think we just only waiting on this right ? Because looks really great for first iteration ! 

Also @portdeveloper could you please resolve conflicts in `packages/nextjs/page/index.tsx`

**Comment by portdeveloper** [comment]
Thanks for reminding me the conflicts!

Here is [the repo for the backend,](https://github.com/portdeveloper/gulltoppr) I use fly.io for deployment so a simple `flyctl launch` is enough for deploying the backend.

**Comment by carletex** [comment]
> Here is [the repo for the backend,](https://github.com/portdeveloper/gulltoppr) I use fly.io for deployment so a simple flyctl launch is enough for deploying the backend.

🫡 Thanks Port! 

I'll do this during the week. I'll hit you up If I have any issues.

---

There is still a conflict in `index.tsx`

**Comment by portdeveloper** [comment]
Sorry about the conflicts 😢 , fixed!

File: packages/nextjs/pages/index.tsx [hunk 2]
```diff
@@ -130,139 +110,170 @@ const Home: NextPage = () => {
...
+      notification.success("ABI successfully loaded.");
+    } catch (error) {
+      notification.error("Invalid ABI format. Please ensure it is a valid JSON.");
+    }
+  };
+
+  const fetchAbiFromHeimdall = async (contractAddress: string) => {
+    setIsFetchingAbi(true);
+    try {
+      const response = await fetch(`https://heimdall-api.fly.dev/${network}/${contractAddress}`);
```
Comment (carletex) [inline packages/nextjs/pages/index.tsx:136]: Can we make this an ENV var?

**Review by carletex — COMMENTED** [review]
Just deployed the BG instance of `gulltoppr` to fly.dev!

https://heimdall-abi-ninja.fly.dev/1/0x6B175474E89094C44Da98b954EedeAC495271d0F

Could you check that everything is working as expected?

Also left a review comment.

**Comment by carletex** [comment]
Just added the envar to Vercel.

When you guys feel that is ready.... feel free to merge!!

**Comment by technophile-04** [comment]
Tested it nicely and working great !!!! 

Unverfied contract on baseSepolia : 0xFB30C0790128b97e3aC540E6124e512E37c47D00 (Heimdall Working)
Unverfied contract on base mainnet : 0x3A6AE1e6aD00518bDcd485a6B4d428fb9c9a3120 (Heimdall not working) 
Unverfied contract on base mainnet : 0xD7c77458e33502f2E5eC04b481463aeCdDd4acd0 (Heimdall working)

Obviously, I think it won't be able to decode each and every contract but its 60% of time able to do so on my random testing 

Will, merge this probably tomorrow 🙌

**Comment by Pabl0cks** [comment]
Been testing it, looks good to me!!

- Is working normally on verified contracts
- Unverified with ABI is still working (can test my sepolia contract `0x60b99996e28c40d5e78ea217e897b93a6fabcdae`  with the ABI on the txt file from my previous comments)
- Unverified with decompile is working quite well, checked with my contract and some random contracts:
   - My sepolia contract `0x60b99996e28c40d5e78ea217e897b93a6fabcdae` is still working nicely
   - Unverified on Base (working): 0x9ba6d87ce0c721e9e7b8d8cac12a5334c8898f59
   - Unverified on Mainnet (+ proxy) (working): 0x758ed0650bdf2ac3bf4a48440c3eb2f6d2bb42a5
   - Unverified on OP mainnet (not working): 0x2e42f214467f647fe687fd9a2bf3baddfa737465
   - Unverified on OP mainnet (working): 0x38edf0a0d32b5c8e23284da87d078062a3f772a7
   - Unverified on OP mainnet (not working): 0x1c8317517070c197ccaaa047662468f1b4055baa

**Pull request merged**